File: composeApp/src/androidMain/kotlin/cafe/cupped/app/designsystem/icons/AppIcon.kt
package cafe.cupped.app.designsystem.icons

import androidx.annotation.DrawableRes
import androidx.compose.foundation.Image
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.res.painterResource
import cafe.cupped.composeapp.R

enum class AppIcon(@DrawableRes val drawableRes: Int, val contentDescription: String) {
    Home(R.drawable.fa_home, "Home"),
    HomeActive(R.drawable.fa_home_active, "Home"),
    Discover(R.drawable.fa_discover, "Discover"),
    DiscoverActive(R.drawable.fa_discover_active, "Discover"),
    Community(R.drawable.fa_community, "Community"),
    CommunityActive(R.drawable.fa_community_active, "Community"),
    Profile(R.drawable.fa_profile, "Profile"),
    ProfileActive(R.drawable.fa_profile_active, "Profile"),
    Add(R.drawable.fa_add, "Add"),
    AddFilled(R.drawable.fa_add_filled, "Add"),
    Mail(R.drawable.fa_mail, "Email"),
    Forward(R.drawable.fa_forward, "Continue"),
    Coffee(R.drawable.fa_coffee, "Coffee"),
    Sparkles(R.drawable.fa_sparkles, "Sparkles"),
    Location(R.drawable.fa_location, "Location"),
    More(R.drawable.fa_more, "More"),
    Favorite(R.drawable.fa_favorite, "Favorite"),
    FavoriteActive(R.drawable.fa_favorite_active, "Favorite"),
    Comment(R.drawable.fa_comment, "Comment"),
    Bookmark(R.drawable.fa_bookmark, "Bookmark"),
    BookmarkActive(R.drawable.fa_bookmark_active, "Bookmark"),
    Box(R.drawable.fa_box, "Box"),
    User(R.drawable.fa_user, "User"),
    Rating(R.drawable.fa_rating, "Rating"),
    Success(R.drawable.fa_success, "Success"),
    Error(R.drawable.fa_error, "Error"),
    Warning(R.drawable.fa_warning, "Warning"),
    Info(R.drawable.fa_info, "Info"),
    Streak(R.drawable.fa_streak, "Streak"),
    Badge(R.drawable.fa_badge, "Badge"),
    Logout(R.drawable.fa_logout, "Sign out"),
}

@Composable
fun AppIcon(
    icon: AppIcon,
    modifier: Modifier = Modifier,
    tint: Color = Color.Unspecified,
    contentDescription: String? = icon.contentDescription,
) {
    Image(
        painter = painterResource(icon.drawableRes),
        contentDescription = contentDescription,
        modifier = modifier,
        colorFilter = tint.takeIf { it != Color.Unspecified }?.let(ColorFilter::tint),
    )
}

File: iosApp/iosApp/DesignSystem/Icons/AppIcon.swift
import SwiftUI

enum AppIcon: String, CaseIterable {
    case home
    case homeActive
    case discover
    case discoverActive
    case community
    case communityActive
    case profile
    case profileActive
    case add
    case addFilled
    case mail
    case forward
    case coffee
    case sparkles
    case location
    case more
    case favorite
    case favoriteActive
    case comment
    case bookmark
    case bookmarkActive
    case box
    case user
    case rating
    case success
    case error
    case warning
    case info
    case streak
    case badge
    case logout

    var assetName: String {
        switch self {
        case .home: "fa_home"
        case .homeActive: "fa_home_active"
        case .discover: "fa_discover"
        case .discoverActive: "fa_discover_active"
        case .community: "fa_community"
        case .communityActive: "fa_community_active"
        case .profile: "fa_profile"
        case .profileActive: "fa_profile_active"
        case .add: "fa_add"
        case .addFilled: "fa_add_filled"
        case .mail: "fa_mail"
        case .forward: "fa_forward"
        case .coffee: "fa_coffee"
        case .sparkles: "fa_sparkles"
        case .location: "fa_location"
        case .more: "fa_more"
        case .favorite: "fa_favorite"
        case .favoriteActive: "fa_favorite_active"
        case .comment: "fa_comment"
        case .bookmark: "fa_bookmark"
        case .bookmarkActive: "fa_bookmark_active"
        case .box: "fa_box"
        case .user: "fa_user"
        case .rating: "fa_rating"
        case .success: "fa_success"
        case .error: "fa_error"
        case .warning: "fa_warning"
        case .info: "fa_info"
        case .streak: "fa_streak"
        case .badge: "fa_badge"
        case .logout: "fa_logout"
        }
    }

    var accessibilityLabel: String {
        switch self {
        case .home, .homeActive: "Home"
        case .discover, .discoverActive: "Discover"
        case .community, .communityActive: "Community"
        case .profile, .profileActive: "Profile"
        case .add, .addFilled: "Add"
        case .mail: "Email"
        case .forward: "Continue"
        case .coffee: "Coffee"
        case .sparkles: "Sparkles"
        case .location: "Location"
        case .more: "More"
        case .favorite, .favoriteActive: "Favorite"
        case .comment: "Comment"
        case .bookmark, .bookmarkActive: "Bookmark"
        case .box: "Box"
        case .user: "User"
        case .rating: "Rating"
        case .success: "Success"
        case .error: "Error"
        case .warning: "Warning"
        case .info: "Info"
        case .streak: "Streak"
        case .badge: "Badge"
        case .logout: "Sign out"
        }
    }
}

struct AppIconView: View {
    let icon: AppIcon
    var size: CGFloat = 20
    var color: Color? = nil
    var decorative: Bool = true

    var body: some View {
        let image = Image(icon.assetName)
            .renderingMode(.template)
            .resizable()
            .scaledToFit()
            .frame(width: size, height: size)

        if decorative {
            image
                .foregroundStyle(color ?? .primary)
                .accessibilityHidden(true)
        } else {
            image
                .foregroundStyle(color ?? .primary)
                .accessibilityLabel(icon.accessibilityLabel)
        }
    }
}
